File: 2023/Leetcode/problem28.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        i = 0
        j = 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                j += 1
                if j == len(needle):
                    return i - len(needle) + 1
            else:
                i -= j
                j = 0

            i += 1
        
        return -1

File: 2024/Leetcode/problem28.ts
// n: haystack.length
// Time Complexity: O(n)
// Space Complexity: O(1)
function strStr(haystack: string, needle: string): number {
    // return haystack.search(needle);
    for (let i = 0; i < haystack.length - needle.length + 1; i++) {
        if (haystack.substring(i, i + needle.length) == needle) {
            return i;
        }
    }
    return -1;
};